dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java

```diff
@@ -405,4 +412,53 @@ public static long countFailures(List<ArtifactResults> table, String columnName)
         .filter(row -> row.getResult(columnName) == null || row.getFailureCount(columnName) > 0)
         .count();
   }
+
+  /**
+   * Returns mapping from jar files to a summary of the root problem in {@link DependencyPath}s of
+   * the jar files. The summary explains common patterns ({@code groupId:artifactId}) in the path
+   * elements. The returned map does not have a key for a jar file when the length of its {@link
+   * DependencyPath} list is small or a common pattern is not found among the items.
+   *
+   * <p>Using this summary in the BOM dashboard avoids repetitive items in the {@link
+   * DependencyPath} list that share the same root problem caused by widely-used libraries, for
+   * example, {@code commons-logging:commons-logging}, {@code
+   * com.google.http-client:google-http-client} and {@code log4j:log4j}.
+   */
+  private static ImmutableMap<String, String> dependencyPathRootProblemMessage(
+      Multimap<Path, DependencyPath> jarToDependencyPaths) {
+    // Freemarker is not good at handling non-string key. Path object in .ftl is automatically
+    // converted to String. https://freemarker.apache.org/docs/app_faq.html#faq_nonstring_keys
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+
+    for (Path jar : jarToDependencyPaths.keySet()) {
+      Set<String> versionlessCoordinates = null; // null for 1st iteration
+      boolean firstIteration = true;
+      ImmutableList<DependencyPath> dependencyPaths = ImmutableList
+          .copyOf(jarToDependencyPaths.get(jar));
+      for (DependencyPath dependencyPath : dependencyPaths) {
+        // Set of versionless coordinates ("groupId:artifactId")
+        Set<String> versionlessCoordinatesInPath =
+            dependencyPath.getPath().stream().map(Artifacts::makeKey).collect(Collectors.toSet());
+        if (firstIteration) {
```

Comment:
can you just create an empty set before starting the loop instead?

Reply:
I find I have to read and really think about this algorithm every time I read it. It's correct so far so far as I can see but not self-evident from the code. If there's nothing simpler, please add some comments explaining how it works.

Reply:
Also, instead of using a boolean firstIteration variable could we instead check that versionlessCoordinates is null or better yet empty?